translate.go

```diff
@@ -4,98 +4,203 @@ import (
-	defaultLocale = l
-}
+	// Format string
+	if s != "" && len(args) > 0 {
+		s = fmt.Sprintf(s, args...)
+	}
 
-// SetLocaleDir sets the directory from which to load locale files
-// only affects GetT() after this call.
-func SetLocaleDir(d string) {
-	fs = http.Dir(d)
+	return s
 }
 
-// SetLocaleFS sets the http.FileStream from from which to load locale files,
-// rather than local directory, only affects GetT() after this call.
-func SetLocaleFS(fs http.FileSystem) {
-	fs = fs
+// Init initializes i18n using the given GetMessagesFn to look up JSON message
+// catalogs for locales, and setting the default locale to the given value.
+// The JSON message catalog is just a map of string keys to string messages. The
+// messages can contain standard Go format strings.
+//
+// If the default locale is not in a valid format, this function will return
+// an error.
+func Init(getMessagesFn GetMessagesFunc, defaultlocale string) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+	getMessages = getMessagesFn
+	var err error
+	defaultLocale, err = newLocale(defaultlocale)
+	if err != nil {
+		return err
+	}
+	m := getMessages(defaultLocale.full)
+	msgsByLang = map[string]messages{defaultLocale.full: m}
+	if !defaultLocale.isLangOnly() {
+		// Also load the language-only resource (if available)
+		msgsByLang[defaultLocale.lang] = getMessages(defaultLocale.lang)
+	}
+
+	// Default current locale to defaultLocale
+	currentLocale = defaultLocale
+	userLocale, err := jibber_jabber.DetectIETF()
+	if err != nil {
+		log.Debugf("Error detecting user locale, defaulting to %v: %v", currentLocale, err)
+	} else if userLocale == "C" {
+		log.Debugf("Unable to detect user locale, defaulting to %v", currentLocale)
+	} else {
+		currentLocale, err = newLocale(userLocale)
+		if err != nil {
+			currentLocale = defaultLocale
+			log.Debugf("Got invalid user locale %v, defaulting to %v: %v", userLocale, currentLocale, err)
+		} else {
+			log.Debugf("Set current locale to %v based on user's locale", currentLocale)
+		}
+	}
+	return nil
 }
 
-type trMap map[string]string
+// InitWithJson is like Init but takes a function that loads JSON data for the
+// messages.
+func InitWithJson(fn func(locale string) ([]byte, error), defaultlocale string) error {
+	return Init(func(locale string) map[string]string {
+		m := make(messages, 0)
+		buf, err := fn(locale)
+		if err != nil {
+			log.Debugf("Error getting locale data for %v: %v", locale, err)
+			return m
+		}
+		if buf == nil {
+			log.Debugf("No locale data found for %v", locale)
+			return m
+		}
+		if err = json.Unmarshal(buf, &m); err != nil {
+			log.Errorf("Error decoding json for locale %v: %v", locale, err)
+		}
+		return m
+	}, defaultlocale)
+}
 
-var fs http.FileSystem
-var defaultLocale string = "en_US"
+// InitWithDir is like Init, but uses resources read from files in the given
+// messagesDir with names corresponding to locales (e.g. en_US).
+func InitWithDir(messagesDir string, defaultlocale string) error {
```

Comment:
@fffw I changed the initialization scheme so that one can now initialize with just a directory.  Above, you'll also see support for just using a function that returns a simple map.  Plus, we still have InitWithJson which takes a function that returns JSON (i.e. tarfs).